### PR TITLE
Add multiflavor support to findbugs

### PR DESCRIPTION
**Pull request by budnyjj**
- create per-flavor findbugs checker tasks
  with automatic classpath setup
- add a list of default android exclusions
  (borrowed from: https://github.com/ChaitanyaPramod/findbugs-android/blob/master/src/main/groovy/com/chaitanyapramod/gradle/android/findbugs/FindBugsAndroidPlugin.groovy)
- generate only one type of findbugs reports per run (prefer xml)

### CODE REVIEW

**Comment by budnyjj** [comment]
@vanniktech , please review my patch, I'll fix tests tomorrow.
Here are some issues with it:
1. Findbugs plugin doesn't allow to create both xml and html report at the same run, so I decided to prefer xml ones. 
2. In android projects source directory could be found automatically, so there is no need to specify 'source' param there.

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 1]
```diff
@@ -10,6 +10,22 @@ import org.gradle.api.tasks.JavaExec
 import org.gradle.api.tasks.Exec
 
 class CodeQualityToolsPlugin implements Plugin<Project> {
+  private static final Collection<String> defaultFindbugsExcludes = [
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: Can you go and take the values from [here](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/blob/681692751f47d64ff3de321b3370e6ea877844e3/src/main/groovy/com/vanniktech/android/junit/jacoco/GenerationPlugin.groovy#L154)?

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 2]
```diff
@@ -159,41 +175,64 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
     return false
   }
 
-  protected static boolean addFindbugs(final Project subProject, final Project rootProject, final CodeQualityToolsPluginExtension extension) {
-    if (!shouldIgnore(subProject, extension) && extension.findbugs.enabled) {
-      final String findbugsClassesPath = isAndroidProject(subProject) ? 'build/intermediates/classes/debug/' : 'build/classes/main/'
+  protected static boolean addFindbugs(final Project subProject, final Project rootProject, final CodeQualityToolsPluginExtension extension
+  ) {
+    if (shouldIgnore(subProject, extension) || !extension.findbugs.enabled) {
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: please revert this change for symmetry reasons

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 3]
```diff
@@ -159,41 +175,64 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
...
+        // findbugs support only one type of report created per run
+        html.enabled = !extension.xmlReports && extension.htmlReports
+        xml.enabled = extension.xmlReports
       }
+      return true
+    }
 
-      subProject.task('findbugs', type: FindBugs, dependsOn: 'assemble') {
-        description = 'Run findbugs'
+    // create a number of findbugs per each flavor and make them dependent on the default one
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 4]
```diff
@@ -305,6 +344,11 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
     return isAndroidLibrary || isAndroidApp
   }
 
+  protected static def getVariants(Project project, Plugin plugin) {
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: do we need the `def`?

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 5]
```diff
@@ -159,41 +175,64 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
...
+    subProject.check.dependsOn defaultTask
+    if (!isAndroidProject(subProject)) {
+      defaultTask.classes = subProject.fileTree('build/classes/main/')
+      // setup provided sources directory for non-android projects
+      defaultTask.source = subProject.fileTree(extension.findbugs.source)
+      defaultTask.classpath = subProject.files()
+
+      defaultTask.reports {
+        // findbugs support only one type of report created per run
+        html.enabled = !extension.xmlReports && extension.htmlReports
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: can you please restore the previous setting here?

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 6]
```diff
@@ -159,41 +175,64 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
...
+    defaultTask.group = 'verification'
+    subProject.check.dependsOn defaultTask
+    if (!isAndroidProject(subProject)) {
+      defaultTask.classes = subProject.fileTree('build/classes/main/')
+      // setup provided sources directory for non-android projects
+      defaultTask.source = subProject.fileTree(extension.findbugs.source)
+      defaultTask.classpath = subProject.files()
+
+      defaultTask.reports {
+        // findbugs support only one type of report created per run
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: remove then

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 7]
```diff
@@ -159,41 +175,64 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
...
-        classes = subProject.fileTree(findbugsClassesPath)
-        source = subProject.fileTree(extension.findbugs.source)
-        classpath = subProject.files()
+        classes = subProject.fileTree(dir:  variant.javaCompile.destinationDir, excludes: defaultFindbugsExcludes)
+        source = variant.sourceSets.java.srcDirs.collect { it.path }.flatten()
+        classpath = variant.javaCompile.classpath
 
         reports {
-          html.enabled = extension.htmlReports
+          // findbugs support only one type of report created per run
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: remove then

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy [hunk 8]
```diff
@@ -159,41 +175,64 @@ class CodeQualityToolsPlugin implements Plugin<Project> {
...
+          html.enabled = !extension.xmlReports && extension.htmlReports
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: can you please restore the previous setting here?

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 1 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: done

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 6 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: done

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 5 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: done

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 3 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: thanks for info! done.

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 7 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: done

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 8 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: done

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 2 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement
http://lecterror.com/articles/view/code-formatting-and-readability

What do you think about it? If you are OK with such style, let me refactor the rest of this class...

Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: If not, I'll revert this change.

Comment (vanniktech) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: I prefer the other way and also changing everything adds too much change to this PR.

**Comment by budnyjj** [comment]
> What about the other tasks do they also need flavor / configuration aware tasks like Findbugs?
Probably, but that's not critical for us now.

**Comment by vanniktech** [comment]
I think it's okay to have this only for findbugs now since all of the other tools don't require the `build/` directory and the bytecode for analyzing.

File: src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy (hunk 2 above)
Comment (budnyjj) [inline src/main/groovy/com/vanniktech/code/quality/tools/CodeQualityToolsPlugin.groovy]: ok

**Comment by budnyjj** [comment]
@vanniktech , please check, if findbugs check work on single-flavored project.

**Comment by budnyjj** [comment]
yes, I'll try, but first we need to define the dependency chains for different types of projects:
1. java project -- findbugs task should depend on compileJava(not assemble since it may require more time to execute)
2. android project without flavors -- findbugs task should depend on compileJava?
3. android project with flavors --
findbugs task should depend on the set of findBugsFlavor tasks, which should depend on corresponding compileJavaFlavor tasks.

@vanniktech , is it correct? How can we test the third case?

**Comment by vanniktech** [comment]
1. If compileJava is enough that's even better, otherwise assemble
2. If compileJava is enough that's even better, otherwise assembleDebug or assembleReleae
3. findbugsMyFlavor should depend on assembleMyFlavor.

**Comment by vanniktech** [comment]
Any updates regarding adding tests for the new behavior?

**Comment by budnyjj** [comment]
I'm going to add them the next week since I'm busy with other activities.

2017-08-10 10:36 GMT+03:00 Niklas Baudy <notifications@github.com>:

**Comment by rock3r** [comment]
Hello there, any news on this? Am I correct in understanding there's only some tests missing?

**Comment by vanniktech** [comment]
I gave this a spin locally but couldn't make it work correctly. Especially adjusting the tests are a pain. 

Furthermore I'm personally not that interested in findbugs anymore since I'm mainly writing Kotlin now and it doesn't work that well. Additionally, it's super slow since it needs bytecode. 

If you ever feel like continuing this please feel free to do so, for now I'll close this PR and the other issues.
